File: 1496-path-crossing/1496-path-crossing.cpp
class Solution {
public:
    bool isPathCrossing(string path) {
        unordered_map<char, pair<int, int>> moves;
        moves['N'] = {0, 1};
        moves['S'] = {0, -1};
        moves['W'] = {-1, 0};
        moves['E'] = {1, 0};
        unordered_set<string> vis;
        vis.insert("0,0");
        int x = 0, y = 0;
        for(char c : path){
            pair<int, int> curr = moves[c];
            x += curr.first;
            y += curr.second;
            string hash = to_string(x) + "," + to_string(y);
            if(vis.find(hash) != vis.end()) return true;
            vis.insert(hash);
        }
        return false;
    }
};

File: 1496-path-crossing/1496-path-crossing.java
class Solution {
    public boolean isPathCrossing(String path) {
        Map<Character, Pair<Integer, Integer>> moves = new HashMap();
        moves.put('N', new Pair(0, 1));
        moves.put('S', new Pair(0, -1));
        moves.put('W', new Pair(-1, 0));
        moves.put('E', new Pair(1, 0));
        Set<Pair<Integer, Integer>> vis = new HashSet();
        vis.add(new Pair(0, 0));
        int x = 0, y = 0;
        for(Character c : path.toCharArray()){
            Pair<Integer, Integer> curr = moves.get(c);
            x += curr.getKey();
            y += curr.getValue();
            Pair<Integer, Integer> pair = new Pair(x, y);
            if(vis.contains(pair)) return true;
            vis.add(pair);
        }
        return false;
    }
}